### PR TITLE
Add Functions for Finding Connected Dominating Sets

### PR DESCRIPTION
**Pull request by GalMichaeli**
## Added Functionality
This PR extends the functionality of
`networkx/networx/algorithms/dominating.py` by adding two functions:

1. `connected_dominating_set` - takes an undirected connected graph `G`
    and returns a connected dominating set (CDS) of G
2. `is_connected_dominating_set` - takes an undirected graph `G` 
    and an iterable `nbunch` and returns true if and only if `nbunch`
    is a CDS of `G`

## Motivation
NetworkX is used extensively in the Networked Software Systems Lab at
the Technion - Israel Institute of Technology during reseach of various
networking topics. Finding a connected dominating set (CDS) in the
context of networks, especially in wireless or ad hoc networks, is
crucial for optimizing communication and resource usage. A CDS serves as
a virtual backbone for the network, meaning efficient routing, power
conservation and network coverage.

## Implementation Details
The function `connected_dominating_set` is an implementation of
Algorithm I as described in [this paper](https://www.scirp.org/reference/referencespapers?referenceid=1979515). Quoting from the paper,
"The idea behind the algorithm is the following: grow a tree T,
starting from the vertex of maximum degree. At each step we pick
a vertex v in T and 'scan it.' Scanning a vertex adds edges to T from v
to all its neighbors not in T. In the end we find a spanning tree T,
and pick the nonleaf nodes as the connected dominating set."

## Testing Details
In order to test `is_connected_dominating_set` we create a path graph G
with four nodes: 0 <-> 1 <-> 2 <-> 3. We then assert that {1, 2} is a
CDS, {1, 3} is NOT a CDS and that {2, 3} indeed induces a connected
subgraph of G but is not a dominating set.
As for `connected_dominating_set`, we first test for the expected behavior
in edge cases, i.e., when G is the null graph, G has only one node and G
is disconnected. Next we test that for complete graphs the function
returns a CDS of size 1. Finally, we define a parametrized test, in which
we test connected Watts-Strogatz graphs with varying parameters and
parametrized seed for reproducibility.

### CODE REVIEW

**Comment by GalMichaeli** [comment]
Hi,
The PR seems to fail due to a missing label.
How can I add the relevant label?

Thanks.

**Comment by dschult** [comment]
I have added the enhancement label -- The labels are assigned by the reviewer rather than the submitter.

Thanks for this! The choice of modules to add look good. I haven't looked at the content yet -- just wanted to get the label in place so you get the "green checkmarks"! :)

**Comment by GalMichaeli** [comment]
Thanks,
Looking forward for your feedback.

File: networkx/algorithms/dominating.py [hunk 1]
```diff
@@ -93,3 +103,262 @@ def is_dominating_set(G, nbunch):
...
+    Raises
+    ------
+    NetworkXNotImplemented
+        If G is directed.
+
+    Notes
+    -----
+    This function implements Algorithm 1 in its basic version as specified
+    in [3]_.
+    Runtime complexity of the algorithm is $O(|E|)$.
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:161]: Paper says the algorithm runs on O(|E|) steps but we need to take into account the time complexity of each step. So it should be O(|E| * log(|N|)).

File: networkx/algorithms/dominating.py [hunk 2]
```diff
@@ -93,3 +103,262 @@ def is_dominating_set(G, nbunch):
...
+    {1, 2, 3, 4, 5, 6, 7}
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If G is directed.
+
+    Notes
+    -----
+    This function implements Algorithm 1 in its basic version as specified
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:159]: We could add the initial paragraph from the paper describing the algorithm

File: networkx/algorithms/dominating.py [hunk 3]
```diff
@@ -93,3 +103,262 @@ def is_dominating_set(G, nbunch):
...
+    # We want a max-heap of the white-degree using heapq, which is a min-heap
+    # So we store the negative of the white-degree
+    gray_nodes = []
+
+    # This will be the CDS
+    black_nodes = set()
+
+    def _update(node):
+        white_nodes.remove(node)
+        push(gray_nodes, (-white_degree[node], node))
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:198]: Having the node in the second position is breaking time complexity. You can use the same trick from Dijkstra's implementation https://github.com/networkx/networkx/blob/main/networkx/algorithms/shortest_paths/weighted.py#L844

File: networkx/algorithms/dominating.py [hunk 4]
```diff
@@ -93,3 +103,262 @@ def is_dominating_set(G, nbunch):
     testset = {n for n in nbunch if n in G}
     nbrs = set(chain.from_iterable(G[n] for n in testset))
     return len(set(G) - testset - nbrs) == 0
+
+
+@not_implemented_for("directed")
+@nx._dispatchable
+def connected_dominating_set(G):
+    r"""Returns a connected dominating set.
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:111]: Should we follow imperative mood? That's probably a question for @dschult

File: networkx/algorithms/dominating.py [hunk 5]
```diff
@@ -93,3 +103,262 @@ def is_dominating_set(G, nbunch):
...
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Dominating_set
+    .. [2] https://en.wikipedia.org/wiki/Connected_dominating_set
+
+    """
+    if nx.is_dominating_set(G, nbunch) == False:
+        return False
+    SG = nx.subgraph(G, nbunch)
+    return nx.is_connected(SG)
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:248]: This can be rewritten as:
```python
return nx.is_dominating_set(G, nbunch) and nx.is_connected(nx.subgraph(G, nbunch))
```

Which makes me think, do we really need this method?

File: networkx/algorithms/dominating.py [hunk 6]
```diff
@@ -68,7 +78,7 @@ def dominating_set(G, start_with=None):
 
 @nx._dispatchable
 def is_dominating_set(G, nbunch):
-    """Checks if `nbunch` is a dominating set for `G`.
+    r"""Checks if `nbunch` is a dominating set for `G`.
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:81]: Why the r change?

File: networkx/algorithms/dominating.py [hunk 7]
```diff
@@ -93,3 +103,262 @@ def is_dominating_set(G, nbunch):
...
+    """
+    if nx.is_dominating_set(G, nbunch) == False:
+        return False
+    SG = nx.subgraph(G, nbunch)
+    return nx.is_connected(SG)
+
+
+@not_implemented_for("directed")
+@nx._dispatchable
+def interval_graph_min_connected_dominating_set(G):
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:253]: Can we leave this in a separate PR so it is easier to review?

File: networkx/algorithms/tests/test_dominating.py [hunk 8]
```diff
@@ -44,3 +44,92 @@ def test_wikipedia_is_dominating_set():
     assert nx.is_dominating_set(G, {4, 3, 5})
     assert nx.is_dominating_set(G, {0, 2})
     assert nx.is_dominating_set(G, {1, 2})
+
```
Comment (amcandio) [inline networkx/algorithms/tests/test_dominating.py:47]: Can you add a suite of tests that generate random graphs and validate the output is a connected dominating set? For example like this one https://github.com/networkx/networkx/pull/7954

**Review by amcandio — COMMENTED** [review]
Sorry for the late review. This is great functionality, left comments!

File: networkx/algorithms/dominating.py (hunk 2 above)
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:159]: Will do!

File: networkx/algorithms/dominating.py (hunk 6 above)
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:81]: Probably the auto-formatter's doing. I don't recall changing this.

File: networkx/algorithms/tests/test_dominating.py (hunk 8 above)
Comment (GalMichaeli) [reply networkx/algorithms/tests/test_dominating.py:47]: Yes. Thanks for the tip.

**Comment by GalMichaeli** [comment]
Minor fix of commit message: the third bullet should be "Rewrite is_connected_dominating_set..."

File: networkx/algorithms/dominating.py [hunk 9]
```diff
@@ -93,3 +102,157 @@ def is_dominating_set(G, nbunch):
...
+    Raises
+    ------
+    NetworkXNotImplemented
+        If G is directed.
+
+    Notes
+    -----
+    This function implements Algorithm 1 in its basic version as described
+    in [3]_. The idea behind the algorithm is the following: grow a tree *T*,
+    starting from the vertex of maximum degree. At each step we pich a vertex
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:160]: Pick*

File: networkx/algorithms/dominating.py [hunk 10]
```diff
@@ -93,3 +102,157 @@ def is_dominating_set(G, nbunch):
...
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Dominating_set
+    .. [2] https://en.wikipedia.org/wiki/Connected_dominating_set
+    .. [3] Guha, S. and Khuller, S.
+           *Approximation Algorithms for Connected Dominating Sets*,
+           Algorithmica, 20, 374-387, 1998.
+
+    """
+    if len(G) == 0 or not nx.is_connected(G):
+        return set()
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:183]: If G is not connected I think algorithm should fail

File: networkx/algorithms/dominating.py [hunk 11]
```diff
@@ -93,3 +102,157 @@ def is_dominating_set(G, nbunch):
...
+        Undirected graph.
+
+    Returns
+    -------
+    black_nodes : set
+        Returns a dominating set of nodes which induce a connected subgraph of G.
+
+    Examples
+    ________
+    >>> G = nx.Graph(
```
Comment (amcandio) [inline networkx/algorithms/dominating.py:129]: Let's add this example to unit tests

File: networkx/algorithms/dominating.py (hunk 6 above)
Comment (amcandio) [reply networkx/algorithms/dominating.py:80]: Let's drop the r and follow imperative style ("Check" instead of "Checks") so we match other docs

File: networkx/algorithms/dominating.py (hunk 4 above)
Comment (amcandio) [reply networkx/algorithms/dominating.py:110]: Let's follow imperative style (see other comment)

File: networkx/algorithms/dominating.py [hunk 12]
```diff
@@ -93,3 +103,262 @@ def is_dominating_set(G, nbunch):
...
+    set *C* that induces a connected subgraph of *G* [2]_.
+
+    Parameters
+    ----------
+    G : NewtorkX graph
+        Undirected graph.
+
+    Returns
+    -------
+    black_nodes : set
```
[the remark this answers is not in the document]
Comment (amcandio) [reply networkx/algorithms/dominating.py:124]: Let's change this line to "connected_dominating_set : set" (it doesn't necessarily need to match variable in the code). This way it will be less confusing for users that don't know about the algorithm details

**Review by amcandio — COMMENTED** [review]
Good stuff. Added minor comments!

File: networkx/algorithms/tests/test_dominating.py [hunk 13]
```diff
@@ -44,3 +44,79 @@ def test_wikipedia_is_dominating_set():
...
+
+def test_is_connected_dominating_set():
```
Comment (amcandio) [inline networkx/algorithms/tests/test_dominating.py:49]: Let's add an assertion for the case of connected but not dominating

File: networkx/algorithms/tests/test_dominating.py [hunk 14]
```diff
@@ -44,3 +44,79 @@ def test_wikipedia_is_dominating_set():
...
+            (9, 12),
+            (10, 12),
+            (11, 12),
+        ]
+    )
+    assert {1, 2, 3, 4, 5, 6, 7} == nx.connected_dominating_set(G)
+
+
+def test_small_connected_watts_strogatz_graph_connected_dominating_set():
+    G = nx.connected_watts_strogatz_graph(10, 3, 0.5)
```
Comment (amcandio) [inline networkx/algorithms/tests/test_dominating.py:108]: Can you fix the seed of this calls in all methods so test case is reproducible? Let's also try with different seeds so we validate function works with more arbitrary graphs. You can also use pytest.parameterize

**Review by amcandio — COMMENTED** [review]
Amazing! Implementation looks good to me, no further comments there. I added other minor comments I missed for the unit tests.

I think we should also polish PR description a bit. Like having different sections:
- What you are adding
- Why you are adding it
- Implementation details. You can also include a brief explanation of the algorithm and quote the paper.
- How you tested it

You can use this one as an example https://github.com/networkx/networkx/pull/7945

**Comment by GalMichaeli** [comment]
> Amazing! Implementation looks good to me, no further comments there. I added other minor comments I missed for the unit tests.
> 
> I think we should also polish PR description a bit. Like having different sections:
> 
>     * What you are adding
> 
>     * Why you are adding it
> 
>     * Implementation details. You can also include a brief explanation of the algorithm and quote the paper.
> 
>     * How you tested it
> 
> 
> You can use this one as an example #7945

Should I edit the original description?

**Comment by amcandio** [comment]
> > Amazing! Implementation looks good to me, no further comments there. I added other minor comments I missed for the unit tests.
> > I think we should also polish PR description a bit. Like having different sections:
> > ```
> > * What you are adding
> > 
> > * Why you are adding it
> > 
> > * Implementation details. You can also include a brief explanation of the algorithm and quote the paper.
> > 
> > * How you tested it
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can use this one as an example #7945
> 
> Should I edit the original description?

Yeah, exactly

File: networkx/algorithms/tests/test_dominating.py (hunk 14 above)
Comment (GalMichaeli) [reply networkx/algorithms/tests/test_dominating.py:108]: Do you mean that I should leave the last three tests (with connected_watts_strogatz_graph) as they are and simply parameterize the seed for the generation for each of them? Or should I merge all tests into one and parameterize the graph generation parameters n, k, p as well as the seed?

Comment (amcandio) [reply networkx/algorithms/tests/test_dominating.py:108]: Yeah, merging them will be best. And then you can parameterize all the variables

Comment (GalMichaeli) [reply networkx/algorithms/tests/test_dominating.py]: I did as you suggested. It really does seem to be the way to go. Thanks.

**Comment by GalMichaeli** [comment]
Please also see the updated PR description.

**Review by amcandio — APPROVED** [review]
Excellent work, approving! @dschult we have a good PR here, can you take a look?

**Review by dschult — COMMENTED** [review]
It looks to me like this finds a spanning tree by repeatedly adding the max-remaining-degree node from all nodes connected to the existing tree.

I would have named white/gray/black as unseen, seen, dominating or something similar which provides a description of what they represent. But I'll leave that up to you.

It often amazes me how heaps end up being so efficient. :)  Here, we have outdated information on the heap -- and we still update it in a lazy way that is efficient in terms of number of updates performed and always correct. On first read, it seems like a lot of updates could be needed, but further thought indicates that this is indeed a really good way to do it. :)

Thanks for this!

File: networkx/algorithms/dominating.py [hunk 15]
```diff
@@ -93,3 +102,163 @@ def is_dominating_set(G, nbunch):
...
+    Notes
+    -----
+    This function implements Algorithm 1 in its basic version as described
+    in [3]_. The idea behind the algorithm is the following: grow a tree *T*,
+    starting from the vertex of maximum degree. At each step we pick a vertex
+    *v* in *T* with maximal number of white neighbors and "scan it." Scanning
+    a vertex add edges to *T* from *v* to all its neighbors not in *T*. In the
+    end we find a spanning tree *T*, and pick the nonleaf nodes as the connected
+    dominating set. Initially all vertices colored white. When we scan a vertex
+    (color it black), we mark all its neighbors that are not in *T* and and add
```
[the remark this answers is not in the document]
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:168]: After some thought, I think I'll change the names to something similar to your suggestion (and change the paragraph accordingly).

File: networkx/algorithms/dominating.py [hunk 16]
```diff
@@ -93,3 +102,163 @@ def is_dominating_set(G, nbunch):
...
+    if len(G) == 1:
+        return set(G)
+
+    G_succ = G._adj  # For speed-up
+
+    push = heappush
+    pop = heappop
+
+    # Keep track of the number of white nodes adjacent to each vertex
+    white_degree = {v: G.degree[v] for v in G}
```
[the remark this answers is not in the document]
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:200]: Oh wow, Great!

File: networkx/algorithms/dominating.py [hunk 17]
```diff
@@ -93,3 +102,163 @@ def is_dominating_set(G, nbunch):
...
+    if not nx.is_connected(G):
+        raise nx.NetworkXError("G must be a connected graph")
+
+    if len(G) == 1:
+        return set(G)
+
+    G_succ = G._adj  # For speed-up
+
+    push = heappush
+    pop = heappop
```
[the remark this answers is not in the document]
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:197]: No problem. I saw this being done in the shortest paths module so thought this was the way to go.

File: networkx/algorithms/dominating.py [hunk 18]
```diff
@@ -93,3 +102,163 @@ def is_dominating_set(G, nbunch):
...
+    white_nodes = set(G)
+
+    # We want a max-heap of the white-degree using heapq, which is a min-heap
+    # So we store the negative of the white-degree
+    gray_nodes = []
+
+    # This will be the CDS
+    black_nodes = set()
+
+    # Use the count c to avoid comparing nodes (may not be able to)
```
[the remark this answers is not in the document]
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:212]: This was suggested by @amcandio, similarly to what was done in dijkstra_multisource.

From the heapq documentation: "Tuple comparison breaks for (priority, task) pairs if the priorities are equal and the tasks do not have a default comparison order" (in our case the priority is the white-degree). 

The suggested solution in the documentation is "A solution [...] is to store entries as 3-element list including the priority, an entry count, and the task. The entry count serves as a tie-breaker so that two tasks with the same priority are returned in the order they were added." 

So I think we should keep this.

**Comment by GalMichaeli** [comment]
Thanks for the comments and suggestions Dan.
I have updated the PR accordingly.

Gal

File: networkx/algorithms/dominating.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (dschult) [reply networkx/algorithms/dominating.py]: Yes -- I agree that we should keep it. 
But the comment suggests that even this approach "may not be able to" avoid comparing nodes.  Maybe just remove the parentheses?

File: networkx/algorithms/dominating.py (hunk 4 above)
Comment (dschult) [reply networkx/algorithms/dominating.py]: It looks like the other existing functions in this module do not use imperative mood.
Library wide there are many `Returns ....` or `Computes ...` or `Checks ...`.
Probably more than the imperative form, but it is also likely to be about half-half. 
We should probably decide about that choice in a community meeting.
Meanwhile, let's keep it similar to the surrounding functions in each module to ease the eventual "unification of mood". :)

Comment (GalMichaeli) [reply networkx/algorithms/dominating.py]: OK, no problem.

File: networkx/algorithms/dominating.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py]: Do you mean to remove only the parentheses or the comment inside them as well?

[the remark this answers is not in the document]
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py]: I also think that `count c` will always avoid comparing the nodes, but for some reason in `_dijkstra_multisource` this was mentioned in the same scenario, so I added it.

[the remark this answers is not in the document]
Comment (dschult) [reply networkx/algorithms/dominating.py]: Ahhh...  I figured there was a reason you had it there...  OK.  Let's remove it here, and I will take a hard look at the queue handling in `multi_source_djikstra` and remove it (or figure out what it is meaning).

Thanks!

[the remark this answers is not in the document]
Comment (amcandio) [reply networkx/algorithms/dominating.py]: Do we allow non comparable objects to be used as nodes? Let's say having strings and integers as nodes of the same graph. If that's the case the comment is correct I believe. The idea is that we don't know if we can so we prevent comparing them.

Note that c is also needed to ensure time complexity. Let's say your node is a long string, then comparing by node will make the length of the node names be part of time complexity.

**Comment by GalMichaeli** [comment]
I’ve removed the change to imperative mood and the comment in the parentheses.

Please let me know if there’s anything else that needs changing.

Thanks.

File: networkx/algorithms/dominating.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (dschult) [reply networkx/algorithms/dominating.py]: Yes -- we allow non comparable objects to be used as nodes -- indeed, any hashable can be a node. So nodes do not need to be comparable. Think of the case where the nodes are themselves graph objects. We want to be able to handle strange cases like that.

As for complexity, all complexity statements and arguments restrict the parameters considered -- or said another way, the "constant" in big-O notation always depends on other things. As a simple example, If we know an algorithm is `O(n + m)` then it is also `O(m)`. This is an inherent issue with the big-O notation -- it doesn't way what limits it is taking. Another example is: `O(x^2)` is often used to mean the limit as `x->0` while `O(n^2)` means the limit `n->inf`. 

Certainly avoiding string comparisons can be called an efficiency issue. And then we don't get into the whole issue of "limits as stuff goes to infinity" complexity proofs. :)  But for nodes, avoiding comparisons is first-and-foremost about our choice to support any hashable object as nodes. If an algorithm is much faster for e.g. integer nodes then we should convert to integers within the function and convert back for return values -- but we only do that in a couple of places. Usually we can handle the general case.  Using `c = count()` for queues is an example of how to do that.  :)

File: networkx/algorithms/dominating.py [hunk 19]
```diff
@@ -93,3 +102,154 @@ def is_dominating_set(G, nbunch):
...
+    Notes
+    -----
+    This function implements Algorithm I in its basic version as described
+    in [3]_. The idea behind the algorithm is the following: grow a tree *T*,
+    starting from the node of maximum degree. Nonleaf nodes in *T* are our
+    connected dominating set (CDS), leaf nodes in *T* are marked as seen and all
+    other nodes are marked as unseen. At each step we pick a seen (leaf) node
+    with maximal number of unseen neighbors, add it to the CDS and mark all its
+    neighbors as seen. The algorithm terminates when there are no more unseen
+    nodes.
```
Comment (dschult) [inline networkx/algorithms/dominating.py:168]: I think this paragraph is not quite right in how neighbors are selected. What do you think of the following?
```suggestion
    This function implements Algorithm I in its basic version as described
    in [3]_. The idea behind the algorithm is the following: grow a spanning tree *T*
    starting from a node with maximum degree and add neighbors based on
    maximal connectivity as follows. Track the degree of each node in the subgraph
    induced by nodes not yet in *T*. At each step, add the node with maximal
    subgraph degree of those already "seen", that is, they are a neighbor of
    some node in *T*. While adding the node to *T*, decrement the degree of
    its neighbors and add them to the heap to be considered as the next node
    to add. The process ends when all "seen" neighbors have been added. 
    The connected dominating set (CDS) is all non-leaf nodes in *T*.
```

File: networkx/algorithms/tests/test_dominating.py [hunk 20]
```diff
@@ -44,3 +44,75 @@ def test_wikipedia_is_dominating_set():
...
+        ]
+    )
+    assert {1, 2, 3, 4, 5, 6, 7} == nx.connected_dominating_set(G)
+
+
+@pytest.mark.parametrize("seed", [1, 13, 29])
+@pytest.mark.parametrize(
+    "n,k,p",
+    [pytest.param(10, 3, 0.2), pytest.param(100, 10, 0.7), pytest.param(1000, 50, 0.5)],
+)
```
Comment (dschult) [inline networkx/algorithms/tests/test_dominating.py:114]: The `pytest.param` is not needed (though can be used to add an ID for the test -- not done here). So maybe this:
```suggestion
@pytest.mark.parametrize("n,k,p", [(10, 3, 0.2), (100, 10, 0.7), (1000, 50, 0.5)])
```

**Review by dschult — COMMENTED** [review]
I'm still working on the notes section...  Can you see what you think of this suggestion?

File: networkx/algorithms/dominating.py (hunk 19 above)
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:168]: A couple of notes:
 1. Our "goal" is to move as many nodes as possible from "unseen" to "seen" in each iteration (note that in each iteration only one node can move from "seen" to the CDS). That's why we keep track of the number of "unseen" neighbors every node has. 
 2. A node in the CDS has no neighbors in "unseen", since we moved those neighbors to "seen" when we moved the node from "seen" to the CDS.
 3. T consists of all nodes in the CDS and the nodes in "seen", so the sentence "Track the degree...." in lines 164-165 is inaccurate, since a node in "seen" may have a neighbor in "seen" but that connection isn't reflected in the tracking of the "unseen-degree" (that's why I named it so). I don't think there's any added value in referring to the "subgraph induced by nodes not yet in T".
 
I think that what I wrote reflects more accurately what's going on in the algorithm, unless I'm not understanding correctly what the code I wrote does. Although it would make it clearer to mention the degree-tracking mechanism.
So I would suggest the following:
`This function implements Algorithm I in its basic version as described
in [3]_. The idea behind the algorithm is the following: grow a tree *T*,
starting from a node with maximum degree. Throughout the growing process,
nonleaf nodes in *T* are our connected dominating set (CDS), leaf nodes in *T* 
are marked as "seen" and nodes in G that are not yet in *T* are marked as "unseen".
We maintain a max-heap of all "seen" nodes, and track the number of "unseen" neighbors
for each node. At each step we pop the heap top--a "seen" (leaf) node with maximal
number of "unseen" neighbors, add it to the CDS and mark all its "unseen" neighbors as "seen".
For each one of the newly created "seen" nodes, we also decrement the number of "unseen" neighbors
for all its neighbors.
The algorithm terminates when there are no more unseen nodes.`

Sorry for the formatting, couldn't figure out how to have it look like yours.

Thanks.

File: networkx/algorithms/tests/test_dominating.py (hunk 20 above)
Comment (GalMichaeli) [reply networkx/algorithms/tests/test_dominating.py:114]: Yes, much better.

File: networkx/algorithms/dominating.py (hunk 19 above)
Comment (dschult) [reply networkx/algorithms/dominating.py:168]: I thought the three categories were CDS nodes, seen nodes, unseen nodes. And I thought they were **distinct** sets.  That is, `seen` includes neighbors of CDS nodes that are not in CDS. `unseen` includes nodes that are not neighbors of nodes in CDS. CDS includes nodes in the tree.

I think you are saying that nodes in `seen` are also in the tree. Don't they only get added to the tree when they are popped from the queue?

Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:168]: > I think you are saying that nodes in seen are also in the tree.

Exactly.

> Don't they only get added to the tree when they are popped from the queue?

 They are added to the tree as soon as they are pushed into "seen" and are added to the CDS when they are popped. The tree isn't explicit in the algorithm. The CDS and "seen" serve the purpose of separating leaf and nonleaf nodes of the tree. In each step we want to add as many nodes as possible to the tree and this can only be done by considering the leaf nodes.

Hope that clears things up.

Comment (dschult) [reply networkx/algorithms/dominating.py:168]: Thanks for that!  
I knew we weren't actually creating the tere, but I didn't have the right model in my head for how the sets related to the tree. "seen" nodes are the current leaves in the tree. "unseen" are not in the tree, and CDS are the nodes that dominate the tree.  That way we don't actually construct the tree and then remove the leaves. They are already split apart.

Your suggestion from a couple comments up looks good to me. 

I'll try to make it a "suggestion" by selecting the lines for the suggestion and adding a comment that starts with "```suggestion" on its own line and ends with "```". [see this stackoverflow or search for github suggestion](https://stackoverflow.com/questions/54239733/how-can-i-manually-add-suggestions-in-code-reviews-on-github)

```suggestion
    This function implements Algorithm I in its basic version as described
    in [3]_. The idea behind the algorithm is the following: grow a tree *T*,
    starting from a node with maximum degree. Throughout the growing process,
    nonleaf nodes in *T* are our connected dominating set (CDS), leaf nodes in *T* 
    are marked as "seen" and nodes in G that are not yet in *T* are marked as "unseen".
    We maintain a max-heap of all "seen" nodes, and track the number of "unseen"
    neighbors for each node. At each step we pop the heap top -- a "seen" (leaf) node
    with maximal number of "unseen" neighbors, add it to the CDS and mark all its
    "unseen" neighbors as "seen". For each one of the newly created "seen" nodes,
    we also decrement the number of "unseen" neighbors for all its neighbors.

    The algorithm terminates when there are no more "unseen" nodes.
```

Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:168]: Glad we got that cleared up. 
So should I change the description according to the previous comment and remove the `pytest.param` from the test and push?

Comment (dschult) [reply networkx/algorithms/dominating.py:168]: Yes -- I think that will resolve all existing conversations. :)

Comment (GalMichaeli) [reply networkx/algorithms/dominating.py]: Great. Pushed.
I sincerely appreciate your patience and guidance.
Thanks Dan and Alejandro.

**Review by dschult — COMMENTED** [review]
One more very small change since we are in here with this code anyway.

Can you remove line 8 and line 10 (importing features from nx directly)
and replace where they are being used with `nx.utils.arbitrary_element` or `nx.utils.not_implemented_for`? That will help later when we clean up imports.

I can do that too if you prefer.

This all looks good to me!

**Review by dschult — APPROVED** [review]

**Comment by GalMichaeli** [comment]
> Can you remove line 8 and line 10 (importing features from nx directly)
   and replace where they are being used with nx.utils.arbitrary_element or nx.utils.not_implemented_for? 
   That will help later when we clean up imports.

Done!

File: networkx/algorithms/dominating.py [hunk 21]
```diff
@@ -68,7 +74,7 @@ def dominating_set(G, start_with=None):
 
 @nx._dispatchable
 def is_dominating_set(G, nbunch):
-    """Checks if `nbunch` is a dominating set for `G`.
+    """Check if `nbunch` is a dominating set for `G`.
```
Comment (dschult) [inline networkx/algorithms/dominating.py:77]:
```suggestion
    """Checks if `nbunch` is a dominating set for `G`.
```

File: networkx/algorithms/dominating.py [hunk 22]
```diff
@@ -93,3 +99,158 @@ def is_dominating_set(G, nbunch):
...
+@nx.utils.not_implemented_for("directed")
+@nx._dispatchable
+def connected_dominating_set(G):
+    """Returns a connected dominating set.
+
+    A *dominating set* for a graph *G* with node set *V* is a subset *D* of
+    *V* such that every node not in *D* is adjacent to at least one
+    member of *D* [1]_. A *connected dominating set* is a dominating
+    set *C* that induces a connected subgraph of *G* [2]_.
+
```
Comment (dschult) [inline networkx/algorithms/dominating.py:113]:
```suggestion

    Connected dominating sets are not unique in general. There may be other connected dominating sets.
    
```

File: networkx/algorithms/dominating.py [hunk 23]
```diff
@@ -93,3 +99,158 @@ def is_dominating_set(G, nbunch):
...
+    Parameters
+    ----------
+    G : NetworkX graph
+        Undirected graph.
+
+    nbunch : iterable
+        An iterable of nodes in the graph `G`.
+
```
Comment (dschult) [inline networkx/algorithms/dominating.py:249]: We need a Returns section of the docs here... (And maybe for the `is_dominating_set` doc_string too.)
You could use the [is_connected function doc_string](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.components.is_connected.html) as a model to work from if that helps.

**Review by dschult — COMMENTED** [review]
This looks good. Just a few more suggestions.

And looking forward beyond this PR. Would it be helpful to have a function that yields all the connected dominating sets? We do have a `SpanningTreeIterator` that yields spanning trees in order of size, so we could remove the leaves from those to get CDSs. I don't know if it would be of interest -- or even if minimum CDS is a thing people care about. But it might not be hard to use existing NX tools to build it.

File: networkx/algorithms/dominating.py (hunk 21 above)
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:77]: How did that get there??

File: networkx/algorithms/dominating.py (hunk 22 above)
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:113]: Good point. Thanks.

File: networkx/algorithms/dominating.py (hunk 23 above)
Comment (GalMichaeli) [reply networkx/algorithms/dominating.py:249]: Great, thanks for the link.

**Comment by GalMichaeli** [comment]
> Would it be helpful to have a function that yields all the connected dominating sets? We do have a SpanningTreeIterator that yields spanning trees in order of size, so we could remove the leaves from those to get CDSs.

That's a great question. If the $i$ th spanning tree returned by the iterator is $ST_i$ then removing each one of its leaves yields a CDS. So I think the number of CDSs is in the order of
$$\sum_{i} \text{leaves}((ST_i)$$
I'm not sure how useful this is.

> I don't know if it would be of interest -- or even if minimum CDS is a thing people care about.

Since finding a minimum CDS (MCDS) is NP-hard, this is an active research topic, especially for node/edge weighted graphs. Though here are algorithms to find an MCDS on certain types of graphs (permutation and interval graphs, and probably others).

**Comment by amcandio** [comment]
It would be great to deeply understand the relationship between spanning trees and CDS. We could even leverage the random spanning tree functionality to derive a random minimal CDS.

What's not clear to me is how to guarantee that the dominating sets are draft uniformly.

**Review by dschult — APPROVED** [review]
I can see why this could become exponential quickly if we start removing each leaf from a tree one at a time. And then all the pairs of leaves, etc. But to find a minimum CDS we would want to remove all the leaves. Do we get close to the minimum CDS by looking at all MSTs and removing all the leaves from them?  

Note that the "minimum" for CDS is just the smallest number of nodes, while the "minimum" for MST is the smallest sum of the weighted edges in the tree. So there might not be any relation at all.

**Most desirable**: are we guaranteed to get a MCDS by removing all leaves from all MSTs?  (I suspect not, especially if they are weighted MSTs)
**Pretty good**: are we likely in most cases to get close to a MCDS using this approach? How close? How do you even measure how close a CDS is from a MCDS? 

Probably this PR is better than that approach -- this PR uses degree info to select the nodes in the tree. So that is at least working toward fewer nodes being non-leaf.  :)   Thanks for this discussion. 

I believe this is ready to be merged.

**Comment by amcandio** [comment]
> Note that the "minimum" for CDS is just the smallest number of nodes, while the "minimum" for MST is the smallest sum of the weighted edges in the tree. So there might not be any relation at all.

Correct, that's why I said just "spanning tree". Each ST could induce a minimal (not minimum) CDS once you remove the leaves and some other nodes maybe?

> **Most desirable**: are we guaranteed to get a MCDS by removing all leaves from all MSTs? (I suspect not, especially if they are weighted MSTs) **Pretty good**: are we likely in most cases to get close to a MCDS using this approach? How close? How do you even measure how close a CDS is from a MCDS?

Regarding "pretty good" there's this paper https://arxiv.org/abs/2303.03125#:~:text=For%20an%20m%2Dedge%20connected,dear%20friend%2C%20Professor%20Takao%20Nishizeki. If I understand that correctly, the algorithm finds a spanning tree with at most 2x more leaves than optimal solution. Meaning that you can probably use that algo to induce a CDSs of size S where |G|-|S| is at most 2x from optimum.

> I believe this is ready to be merged.

+1 Good job @GalMichaeli!
